### PR TITLE
Re-index stories in Yoast SEO if permalinks are outdated

### LINKED ISSUE #4754: Yoast SEO: stale cache after permalinks changed
## Bug Description

In 1.0.0 release, we switched the base URL for stories to be at domain.com/web-stories/ instead of the prior domain.com/stories/. However, some stories on sites using Yoast are still pointing to the old /stories/ URL as their canonical, triggering an issue with non-self canonical pages. 

## Expected Behaviour

Canonicals should be updated/fixed to point to the latest /web-stories/ canonical URL. 

## Steps to Reproduce

Examples where this has been seen: 
https://lsv-paf.de/stories/sunrise2020/
https://knaptonwright.co.uk/stories/website-copywriting/
https://wecomm.fr/web-stories/pourquoi-creer-site-web/

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

We changed the rewrite slug from `/stories/` to `/web-stories/` mid-way.

Because Yoast SEO caches the permalinks in the database, it means stories created prior to that change will have outdated canonical URLs and the like

## Relevant Technical Choices

Usage of all these Yoast SEO classes required PHPStan stubs. Took the time to improve the PHPStan config using the recommendations at https://phpstan.org/user-guide/discovering-symbols.

Unfortunately https://github.com/php-stubs/wordpress-seo-stubs is outdated, otherwise we could have used that package directly for the stubs.

## To-do

N/A

## User-facing changes

N/A

## Testing Instructions

1. Activate Yoast SEO
1. Install Web Stories for WordPress Beta 2
1. Publish a story
1. Update to 1.0.0 stable
1. Check story on the frontend, notice the **old** URLs
1. Switch to this branch
1. Visit WordPress admin to trigger the db upgrader
1. Check story on the frontend, notice the **new** URLs

---

Fixes #4754

### CODE REVIEW

File: includes/Database_Upgrader.php [hunk 1]
```diff
@@ -26,6 +26,8 @@
 
 namespace Google\Web_Stories;
 
+use Yoast\WP\SEO\Actions\Indexation\Indexable_Prepare_Indexation_Action;
+
```
Comment (swissspidy) [inline includes/Database_Upgrader.php]:
```suggestion
```

**Comment by swissspidy** [comment]
@omarreiss friendly ping 👋 curious to hear from an expert whether the approach we're taking here makes sense. Would love to get your input!

**Comment by herregroen** [comment]
@swissspidy Hey, this approach would definitely work and if the expected use-case is always a relatively small number of stories total ( say below 100 ) then you can keep this.

If you're expecting sites to exist that have potentially thousands of web stories then it'll become likely that rebuilding all those is going to take more time and thus either significantly slow down the update request or even timeout.

For those cases we've build in logic where you can set the permalink column to `NULL` in which case it will automatically be recalculated for every object as it's fetched.

You could use a query like `$indexable_repository->query()->where( 'object_type', 'post' )->where( 'object_sub_type', Story_Post_Type::POST_TYPE_SLUG )->set( 'permalink', null )->update_many();` to achieve this.

I believe 15.2 will also introduce a helper function to achieve resetting permalinks but that may not be feasible for your release timeline.

**Comment by swissspidy** [comment]
Thanks @herregroen, that's very helpful!

I expect the number of affected stories to be very low as the plugin is still very new, so it should be fine to keep. However, the set-permalink-to-`NULL` approach is very intriguing due to its simplicity. I might take a stab at doing that instead.

File: includes/Database_Upgrader.php [hunk 2]
```diff
@@ -274,6 +275,59 @@ protected function rewrite_flush() {
...
+		/**
+		 * Indexable Builder.
+		 *
+		 * @var \Yoast\WP\SEO\Builders\Indexable_Builder $builder Indexable Builder.
+		 */
+		$builder = YoastSEO()->classes->get( 'Yoast\WP\SEO\Builders\Indexable_Builder' );
+
+		$indexable_before = $repository->find_for_post_type_archive( Story_Post_Type::POST_TYPE_SLUG, false );
+
+		if ( ! $indexable_before || false === strpos( $indexable_before->permalink, '/web-stories//' ) ) {
```
Comment (spacedmonkey) [inline includes/Database_Upgrader.php]: Why does this line have two slashes in it?

File: includes/Database_Upgrader.php [hunk 3]
```diff
@@ -274,6 +275,59 @@ protected function rewrite_flush() {
...
+			$builder->build_for_post_type_archive( Story_Post_Type::POST_TYPE_SLUG, $indexable_before );
+		}
+
+		$all_stories = get_posts(
+			[
+				'fields'           => 'ids',
+				'suppress_filters' => false,
+				'post_type'        => [ Story_Post_Type::POST_TYPE_SLUG ],
+				'posts_per_page'   => - 1,
```
Comment (spacedmonkey) [inline includes/Database_Upgrader.php]: Platform like VIP go, limit requests to 100, even if you request -1. See this sniff. 
https://github.com/Automattic/VIP-Coding-Standards/blob/bfc7fc56e18bec55a3830de06d1d3be0443585dd/WordPress-VIP-Go/ruleset.xml#L142-L144

Comment (spacedmonkey) [inline includes/Database_Upgrader.php]:
```suggestion
				'update_post_meta_cache' => false,
				'update_post_term_cache'  => false,
				'posts_per_page'   => - 1,
```

No need to load meta caches here.

Comment (spacedmonkey) [reply includes/Database_Upgrader.php]: Limiting this query is basically what @herregroen said. Maybe a while loop of 100 chunk. I would be EXTREMELY surprised if there are sites with more than 100 stories in the wild, but you never know.

File: includes/Database_Upgrader.php (hunk 2 above)
Comment (swissspidy) [reply includes/Database_Upgrader.php]: Typo

File: includes/Database_Upgrader.php (hunk 3 above)
Comment (swissspidy) [reply includes/Database_Upgrader.php]: FWIW I copied this from `includes/namespace.php`. Do we need to change that there as well?

File: includes/Database_Upgrader.php (hunk 2 above)
Comment (swissspidy) [inline includes/Database_Upgrader.php]:
```suggestion
		if ( ! $indexable_before || false === strpos( $indexable_before->permalink, '/web-stories/' ) ) {
```

File: includes/Database_Upgrader.php (hunk 3 above)
Comment (swissspidy) [reply includes/Database_Upgrader.php]: I would be EXTREMELY surprised as well, that's why I went with this. I can change it to 100 but I don't think we need pagination / a loop.

Comment (spacedmonkey) [reply includes/Database_Upgrader.php]: There is a different here, this all you want to do is get a list of IDs for posts to update them. In that other place, you want to warm caches into memory before deleting.

**Comment by spacedmonkey** [comment]
This is some small nit pic, but this is good to go.

**Comment by swissspidy** [comment]
@spacedmonkey mind re-reviewing this one?

**Comment by swissspidy** [comment]
@spacedmonkey friendly ping

**Review by spacedmonkey — APPROVED** [review]
Sorry, i thought I had already approved this. 👍

**Pull request merged**